Libraries/LibCrypto/Curves/SECPxxxr1.h

```diff
@@ -363,72 +364,54 @@ class SECPxxxr1 : public EllipticCurve {
 
-        StorageType u1 = modular_multiply_order(z_mo, s_inv);
-        StorageType u2 = modular_multiply_order(r_mo, s_inv);
+        auto* params = OPENSSL_TRY_PTR(OSSL_PARAM_BLD_to_param(params_bld));
+        ScopeGuard const free_params = [&] { OSSL_PARAM_free(params); };
 
-        u1 = from_montgomery_order(u1);
-        u2 = from_montgomery_order(u2);
+        auto key = TRY(OpenSSL_PKEY::wrap(EVP_PKEY_new()));
+        auto* key_ptr = key.ptr();
+        OPENSSL_TRY(EVP_PKEY_fromdata(ctx_import.ptr(), &key_ptr, EVP_PKEY_PUBLIC_KEY, params));
 
-        JacobianPoint point1 = TRY(generate_public_key_internal(u1));
-        JacobianPoint point2 = TRY(compute_coordinate_internal(u2, JacobianPoint {
-                                                                       unsigned_big_integer_to_storage_type(pubkey.x),
-                                                                       unsigned_big_integer_to_storage_type(pubkey.y),
-                                                                       1u,
-                                                                   }));
+        auto ctx = TRY(OpenSSL_PKEY_CTX::wrap(EVP_PKEY_CTX_new_from_pkey(nullptr, key.ptr(), nullptr)));
 
-        // Convert the input point into Montgomery form
-        point1.x = to_montgomery(point1.x);
-        point1.y = to_montgomery(point1.y);
-        point1.z = to_montgomery(point1.z);
+        OPENSSL_TRY(EVP_PKEY_verify_init(ctx.ptr()));
 
-        VERIFY(is_point_on_curve(point1));
+        auto* sig_obj = OPENSSL_TRY_PTR(ECDSA_SIG_new());
+        ScopeGuard const free_sig_obj = [&] { ECDSA_SIG_free(sig_obj); };
 
-        // Convert the input point into Montgomery form
-        point2.x = to_montgomery(point2.x);
-        point2.y = to_montgomery(point2.y);
-        point2.z = to_montgomery(point2.z);
+        auto r = TRY(unsigned_big_integer_to_openssl_bignum(signature.r));
+        auto s = TRY(unsigned_big_integer_to_openssl_bignum(signature.s));
 
-        VERIFY(is_point_on_curve(point2));
+        // Let sig_obj own a copy of r and s
+        OPENSSL_TRY(ECDSA_SIG_set0(sig_obj, BN_dup(r.ptr()), BN_dup(s.ptr())));
 
-        JacobianPoint result = point_add(point1, point2);
+        u8* sig = nullptr;
+        ScopeGuard const free_sig = [&] { OPENSSL_free(sig); };
 
-        // Convert from Jacobian coordinates back to Affine coordinates
-        convert_jacobian_to_affine(result);
-
-        // Make sure the resulting point is on the curve
-        VERIFY(is_point_on_curve(result));
-
-        // Convert the result back from Montgomery form
-        result.x = from_montgomery(result.x);
-        result.y = from_montgomery(result.y);
-        // Final modular reduction on the coordinates
-        result.x = modular_reduce(result.x);
-        result.y = modular_reduce(result.y);
+        auto sig_len = i2d_ECDSA_SIG(sig_obj, &sig);
+        if (sig_len <= 0) {
+            OPENSSL_TRY(sig_len);
```

Comment:
This is a bit...confusing, maybe something like
```suggestion
        auto sig_len = [&] {
            auto ret = i2d_ECDSA_SIG(sig_obj, &sig);
            if (ret <= 0) {
                OPENSSL_TRY(ret);
                VERIFY_NOT_REACHED();
            }
            return ret;
        }();
```
could make it clearer? or just renaming `sig_len` to `sig_len_ret`? 🤔

Reply:
Refactored using a lambda :+1: